Stratis.Bitcoin.Features.Wallet/WalletManager.cs

```diff
@@ -639,63 +640,70 @@ public void ProcessBlock(Block block, ChainedBlock chainedBlock)
                 this.UpdateLastBlockSyncedHeight(chainedBlock);
             }
         }
-
+        
         /// <inheritdoc />
         public void ProcessTransaction(Transaction transaction, int? blockHeight = null, Block block = null)
         {
             Guard.NotNull(transaction, nameof(transaction));
 
             var hash = transaction.GetHash();
-            this.logger.LogTrace($"transaction received - hash: {hash}, coin: {this.coinType}");
+            this.logger.LogTrace("transaction received - hash: {0}, coin: {1}", hash, this.coinType);
 
             // load the keys for lookup if they are not loaded yet.
             if (this.keysLookup == null)
             {
                 this.LoadKeysLookupLock();
             }
-
-            var foundtrx = new List<Tuple<Script, uint256>>();
+            
+            var foundtrx = new HashSet<(Script Script, uint256 TxHash)>();
 
             lock (this.lockObject)
             {
                 // check the outputs
-                foreach (TxOut utxo in transaction.Outputs)
+                foreach (TxOut txo in transaction.Outputs)
                 {
                     // check if the outputs contain one of our addresses
-                    if (this.keysLookup.TryGetValue(utxo.ScriptPubKey, out HdAddress pubKey))
+                    foreach (var script in this.keysLookup.Keys)
```

Comment:
This is not good as you don't leverage the hash lookup.
you need to use `keysLookup.TryGetValue`